src/terrain/util/keycloak_oidc.clj

```diff
@@ -7,24 +7,44 @@
             [terrain.util.config :as config]
             [otel.otel :as otel]))
 
+(defn- is-service-account?
+  [{:keys [preferred_username]}]
+  (string/starts-with? preferred_username "service-account-"))
+
 (defn user-from-token
   "Extracts user information from a Keycloak OIDC token."
   [{:keys [jwt-claims]}]
-  {:shortUsername (:preferred_username jwt-claims)
-   :username      (str (:preferred_username jwt-claims) "@" (config/uid-domain))
-   :email         (:email jwt-claims)
-   :firstName     (:given_name jwt-claims)
-   :lastName      (:family_name jwt-claims)
-   :commonName    (:name jwt-claims)
-   :entitlement   (:entitlement jwt-claims)})
+  (if (is-service-account? jwt-claims)
+    nil
+    {:shortUsername (:preferred_username jwt-claims)
+     :username      (str (:preferred_username jwt-claims) "@" (config/uid-domain))
+     :email         (:email jwt-claims)
+     :firstName     (:given_name jwt-claims)
+     :lastName      (:family_name jwt-claims)
+     :commonName    (:name jwt-claims)
+     :entitlement   (:entitlement jwt-claims)}))
+
+(defn service-account-from-token
+  "Extract service account information from a Keycloak OIDC token."
+  [{:keys [jwt-claims]}]
+  (if (is-service-account? jwt-claims)
+    {:username (:preferred_username jwt-claims)
+     :roles (:roles (:realm_access jwt-claims))}
+    nil))
 
 (def ^:private required-claims
   [:preferred_username :email :given_name :family_name :name :entitlement])
 
+(def ^:private required-service-account-claims
+  [:preferred_username :entitlement :realm_access])
```

Comment:
I think that `:entitlement` can be removed from the list of required claims for service accounts. This field is present because of the way that Keycloak is configured, but chances are that it will always be empty for service accounts.

Reply:
Good to know. I'd been wondering about it and figured I'd keep it since it was there and in the older `required-claims`, but I'll remove it since it's probably useless (and after all, we're not using it).

Thanks!